tensorflow_encrypted/protocol/pond.py

```diff
@@ -120,6 +116,51 @@ def define_private_variable(self, initial_value, apply_encoding=True, name=None)
         _initializers.append(x.initializer)
         return x
 
+    def define_public_input(self, provider:InputProvider, apply_scaling:bool=True, name:str=None) -> 'PondPublicTensor':
+
+        with tf.name_scope('public-input{}'.format('-'+name if name else '')):
+
+            with tf.device(provider.player.device_name):
+
+                v = provider.provide_input()
+                v = _encode(v, apply_scaling)
+                x_on_0 = v
+                x_on_1 = v
+
+        x = PondPublicTensor(self, x_on_0, x_on_1)
+        return x
+
+    def define_private_input(self, provider:InputProvider, apply_scaling:bool=True, name:str=None) -> 'PondPrivateTensor':
+
+        with tf.name_scope('private-input{}'.format('-'+name if name else '')):
+
+            with tf.device(provider.player.device_name):
+
+                v = provider.provide_input()
+                v = _encode(v, apply_scaling)
+                x0, x1 = _share(v)
+
+        x = PondPrivateTensor(self, x0, x1)
+        return x
+
+    def define_output(self, x, receiver:OutputReceiver, apply_scaling=True, name=None):
+        assert isinstance(x, PondPrivateTensor), type(x)
+
+        x0, x1 = x.unwrapped
+
+        with tf.name_scope('output{}'.format('-'+name if name else '')):
+
+            with tf.device(receiver.player.device_name):
+
+                v: BackingTensor = _reconstruct(x0, x1)
+                # v = _decode(v, apply_scaling) # TODO[Morten] re-enable
+                op = receiver.receive_output(v.backing[0])
+
+                # wrap in tf.group to prevent sending back any tensors (which might hence be leaked)
+                op = tf.group(op)
```

Comment:
Interesting, why does `tf.group` avoid sending back tensors?

Reply:
I can't find the link I initially used, but you essentially want something that returns an tf.Operation instead of tf.Tensor -- [tf.group](https://www.tensorflow.org/api_docs/python/tf/group) does that

Reply:
In this case, maybe put this line inside `receive_output `?

Reply:
as you also suggested, wanted to enforce this one and hence put it in the protocol